src/codegen/compile.ml

```diff
@@ -1574,58 +1668,223 @@ module BitTagged = struct
     compile_bitand_const 0x1l ^^
     E.if_ env retty is2 is1
 
-  (* 64 bit numbers *)
+  let ubits_of pty = TaggingScheme.ubits_of pty
 
-  (* static *)
-  let can_tag_const (n : int64) =
-    let lower_bound = Int64.(neg (shift_left 1L 30)) in
-    let upper_bound = Int64.shift_left 1L 30 in
-    lower_bound <= n && n < upper_bound
+  let sbits_of pty = (ubits_of pty) - 1
 
-  let tag_const i = Int32.shift_left (Int64.to_int32 i) 1
+  (* 64 bit numbers *)
 
+  (* static *)
+  let can_tag_const pty (n : int64) = Type.(
+    match pty with
+    |  Nat | Int | Int64 | Int32 ->
+      let sbits = sbits_of pty in
+      let lower_bound = Int64.(neg (shift_left 1L sbits)) in
+      let upper_bound = Int64.shift_left 1L sbits in
+      lower_bound <= n && n < upper_bound
+    | Nat64 | Nat32 ->
+      let ubits = ubits_of pty in
+      let upper_bound = Int64.shift_left 1L ubits in
+      0L <= n && n < upper_bound
+    | _ -> assert false)
+
+  let tag_const pty i = Type.(
+    match pty with
+    |  Nat | Int | Int64 | Int32 ->
+      Int32.shift_left (Int64.to_int32 i) (32 - ubits_of pty)
+      (* tag *)
+      |> Int32.logor (TaggingScheme.tag_of_typ pty)
+    |  Nat64 | Nat32 ->
+      Int32.shift_left (Int64.to_int32 i) (32 - ubits_of pty) (* TBR *)
+      (* tag *)
+      |> Int32.logor (TaggingScheme.tag_of_typ pty)
```

Comment:
Maybe a brief not could be helpful (if I understood correctly):
For signed small (=non-boxed) numbers, the bit at position `ubits_of pty - 1` serves as the sign bit. 
Therefore, the tagging logic is the same for signed and unsigned small numbers.

```suggestion
    |  Nat | Int | Int64 | Int32 | Nat64 | Nat32 ->
      Int32.shift_left (Int64.to_int32 i) (32 - ubits_of pty)
      (* tag *)
      |> Int32.logor (TaggingScheme.tag_of_typ pty)
```